### ISSUE TITLE
terraform plan -refresh-only -detailed-exitcode gives incorrect exit code 2 even if the console output says no changes detected

### ISSUE THREAD

**Issue by shaik262** [hashicorp/terraform#35117]
### Terraform Version

```shell
v1.8.1
```

### Terraform Configuration Files

```terraform
...terraform config...
```

### Debug Output

I tried to get the exit code displayed on console it gives 2 even if there are no changes detected as per terraform

### Expected Behavior

It should return 0 as there are no changes present as per the console

### Actual Behavior

terraform plan -refresh-only -detailed-exitcode gives incorrect exit code 2 even if the console output says no changes detected

### Steps to Reproduce

we used this command as a part of ADO pipeline drift detection 
$terraform plan -refresh-only -detailied-exitcode

### Additional Context

![image](https://github.com/hashicorp/terraform/assets/169045931/dd6a7527-12c7-45c0-ac76-9f38138149f0)
![Screenshot 2024-05-06 160231](https://github.com/hashicorp/terraform/assets/169045931/8bf46280-e38c-40b4-ba5b-17b34d175f1d)

**Comment by jbardin** [comment]
Hi @shaik262,

Thanks for filing the issue! I'm not able to replicate the result you are showing here, and seeing a `0` exit code for empty refresh plans. Can you show exactly how you are setting up and executing `terraform` to get this output?

Thanks!

**Comment by shaik262** [comment]
Hi jbardin thank you for looking into this issue. 

We have an azure pipeline which has a drift detection in place (cronjob based) This schedule phase uses a command to check the drif and depending on the exit code it goes to review and apply stage. Here is the logic 
We are using 
The agent we are using to run this is ubuntu. 


![35EF6CB6-AC17-4AA1-8EED-FB01280C0BC5](https://github.com/hashicorp/terraform/assets/169045931/1759655d-1168-4938-9c09-b3861530674b)


![CB4D1DB9-77CB-4C1E-A444-3D4FCF8A4793](https://github.com/hashicorp/terraform/assets/169045931/5eb351cd-4874-4fac-813d-6495fbac1917)

**Comment by shaik262** [comment]
[terraform.txt](https://github.com/hashicorp/terraform/files/15233277/terraform.txt)
[azure-pipelines.txt](https://github.com/hashicorp/terraform/files/15233286/azure-pipelines.txt)

**Comment by jbardin** [comment]
Thanks @shaik262. That unfortunately doesn't give me enough to reproduce the behavior. I think what we need is the plan data or Terraform configuration rather than the pipeline configuration. 

Have you applied a plan to this configuration since upgrading to v1.8? See the release notes or [upgrade guide](https://developer.hashicorp.com/terraform/language/upgrade-guides#possible-spurious-changes-when-refreshing) about using `-refresh-only` after upgrading.

**Comment by shaik262** [comment]
Hi @jbardin I went trough the link which you've sent and applied the planned changes, but again in the next schedule the pipeline is waiting in review and apply stage even though there are no changes being displayed by terraform plan. ( the detailed exit code is "2")

**Comment by shaik262** [comment]
Hello @jbardin. We have resolved this issue. The Terraform Plan show command for the file with resource drift is giving an output which has a property relevant_attributes,  resource_drift and check along with other properties. But the file which has no changes, is giving a output without the relevant_attribute and resources_drift. We changed our condition to look for just resources_drift in the output of terraform plan show  for any changes as this property is generated in plan to let us know which external changes are  causing this drift.

Issue : For some of the files with no changes, the same command is generating relevant_attributes also which lists the sources of all values contributing to changes in the plan.

 
![image](https://github.com/hashicorp/terraform/assets/169045931/01b31775-255c-4c54-aede-25d64b65d5e6)
